File: python/wdw.py
# -*- coding: utf-8 -*-
"""
WDW 大文本数据压缩 API SDK (Python)
====================================

官方端点: https://www.zhongbaomaoyi.cn/fwq/shujuyasuo/compress_api.php

用法:
    from wdw import WdwClient
    c = WdwClient(key="你的API Key")
    b85 = c.compress("要压缩的文本")          # 返回 <~...~> 形式的 base85 文本
    text = c.decompress(b85)                 # 还原原文

    # 二进制模式 (体积最小)
    raw = c.compress_bytes(b"二进制数据")     # 返回 bytes
    back = c.decompress_bytes(raw)

依赖: 仅 Python 标准库 (urllib)，无需 pip install。
"""

from __future__ import annotations

import base64
import json
import urllib.parse
import urllib.request
from typing import Optional


class WdwError(Exception):
    """服务器返回的业务错误 (code != 1 / HTTP body 是 JSON 错误体)。"""

    def __init__(self, msg: str, http_status: int = 200):
        super().__init__(msg)
        self.http_status = http_status


class WdwClient:
    DEFAULT_ENDPOINT = "https://www.zhongbaomaoyi.cn/fwq/shujuyasuo/compress_api.php"

    def __init__(
        self,
        key: str,
        endpoint: str = DEFAULT_ENDPOINT,
        device_id: Optional[str] = None,
        timeout: float = 30.0,
    ):
        """
        :param key:        API Key (登录后在控制台获取；游客模式可留空但额度极小)
        :param endpoint:   接口地址，一般不用改
        :param device_id:  游客模式设备 ID；登录用户可忽略
        :param timeout:    超时秒数
        """
        self.key = key
        self.endpoint = endpoint
        self.device_id = device_id
        self.timeout = timeout

    # ---------- 内部：统一请求 ----------
    def _post(self, form: dict, expect_binary: bool) -> object:
        data = urllib.parse.urlencode(form).encode("utf-8")
        req = urllib.request.Request(
            self.endpoint,
            data=data,
            headers={
                "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
                "User-Agent": "wdw-sdk-python/1.0",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body = resp.read()
                status = resp.status
        except urllib.error.HTTPError as e:
            body = e.read()
            status = e.code

        # 服务器错误时返回 JSON: {"code":0,"msg":"..."}；成功时是裸文本或裸二进制
        stripped = body.lstrip()
        if stripped[:1] == b"{":
            try:
                obj = json.loads(body.decode("utf-8", errors="replace"))
                if isinstance(obj, dict) and "msg" in obj:
                    raise WdwError(obj.get("msg", "未知错误"), status)
            except json.JSONDecodeError:
                pass
        if expect_binary:
            return body
        return body.decode("utf-8", errors="replace")

    def _base_form(self, lx: int, fhlx: int) -> dict:
        form = {"lx": str(lx), "fhlx": str(fhlx), "key": self.key}
        if self.device_id:
            form["device_id"] = self.device_id
        return form

    # ---------- 对外：文本模式 (fhlx=1) ----------
    def compress(self, text: str) -> str:
        """压缩一段 UTF-8 文本，返回带 <~...~> 包装的 base85 字符串。"""
        form = self._base_form(lx=1, fhlx=1)
        form["clwb"] = text
        return self._post(form, expect_binary=False)

    def decompress(self, b85: str) -> str:
        """传入 compress() 返回的完整 <~...~> 字符串，还原原文。"""
        form = self._base_form(lx=2, fhlx=1)
        form["clwb"] = b85
        return self._post(form, expect_binary=False)

    # ---------- 对外：二进制模式 (fhlx=2) ----------
    def compress_bytes(self, raw: bytes) -> bytes:
        """压缩 bytes，返回压缩后的裸二进制 (.wdw)。"""
        # fhlx=2 压缩时，clwb 仍走表单 (文本)；这里用 latin-1 透传任意字节
        form = self._base_form(lx=1, fhlx=2)
        form["clwb"] = raw.decode("latin-1")
        return self._post(form, expect_binary=True)

    def decompress_bytes(self, raw: bytes) -> str:
        """解压 .wdw 二进制，返回原文 (str)。"""
        # fhlx=2 解压：原始二进制 body 直接 POST，参数走 query string
        qs = urllib.parse.urlencode({
            "lx": 2, "fhlx": 2, "key": self.key,
            **({"device_id": self.device_id} if self.device_id else {}),
        })
        req = urllib.request.Request(
            f"{self.endpoint}?{qs}",
            data=raw,
            headers={
                "Content-Type": "application/octet-stream",
                "User-Agent": "wdw-sdk-python/1.0",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body = resp.read()
                status = resp.status
        except urllib.error.HTTPError as e:
            body = e.read()
            status = e.code
        if body[:1] == b"{":
            try:
                obj = json.loads(body.decode("utf-8", errors="replace"))
                raise WdwError(obj.get("msg", "未知错误"), status)
            except json.JSONDecodeError:
                pass
        return body.decode("utf-8", errors="replace")


# ---------- 本地 base85 <-> bytes 工具 (可选，不用调服务器) ----------
# 官方字母表 (RFC 1924)，与 <~...~> 包装配套
def b85_to_bytes(b85: str) -> bytes:
    """把服务器返回的 <~...~> base85 文本解码成原始 bytes。"""
    s = b85.strip()
    if s.startswith("<~"):
        s = s[2:]
    if s.endswith("~>"):
        s = s[:-2]
    return base64.b85decode(s)


def bytes_to_b85(data: bytes) -> str:
    """把任意 bytes 编码成带 <~...~> 包装的 base85 文本。"""
    return "<~" + base64.b85encode(data).decode("ascii") + "~>"


# ---------- CLI 入口 ----------
if __name__ == "__main__":
    import sys

    if len(sys.argv) < 3:
        print("用法: python wdw.py <API_KEY> <compress|decompress> [输入文件]")
        sys.exit(1)
    key, op = sys.argv[1], sys.argv[2]
    cli = WdwClient(key=key)
    if op == "compress":
        text = sys.stdin.read() if len(sys.argv) < 4 else open(sys.argv[4], encoding="utf-8").read()
        print(cli.compress(text))
    elif op == "decompress":
        b85 = sys.stdin.read() if len(sys.argv) < 4 else open(sys.argv[4], encoding="utf-8").read()
        print(cli.decompress(b85.strip()))
    else:
        print("未知操作:", op)
        sys.exit(1)

File: go/wdw.go
// Package wdw 是 WDW 大文本数据压缩 API 的官方开源 Go SDK。
//
// 端点: https://www.zhongbaomaoyi.cn/fwq/shujuyasuo/compress_api.php
//
// 用法:
//   client := wdw.New("你的API_KEY")
//   b85, _ := client.Compress("要压缩的文本")
//   text, _ := client.Decompress(b85)
package wdw

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const DefaultEndpoint = "https://www.zhongbaomaoyi.cn/fwq/shujuyasuo/compress_api.php"

// Client WDW API 客户端
type Client struct {
	Key      string
	Endpoint string
	DeviceID string
	Timeout  time.Duration
	HTTPClient *http.Client
}

// New 创建一个客户端
func New(key string) *Client {
	return &Client{
		Key:      key,
		Endpoint: DefaultEndpoint,
		Timeout:  30 * time.Second,
		HTTPClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// Compress 压缩 UTF-8 文本，返回 <~...~> 包装的 base85 字符串
func (c *Client) Compress(text string) (string, error) {
	form := url.Values{}
	form.Set("clwb", text)
	form.Set("lx", "1")
	form.Set("fhlx", "1")
	return c.postForm(form, false)
}

// Decompress 传入 Compress 返回的完整字符串，还原原文
func (c *Client) Decompress(b85 string) (string, error) {
	form := url.Values{}
	form.Set("clwb", b85)
	form.Set("lx", "2")
	form.Set("fhlx", "1")
	return c.postForm(form, false)
}

// CompressBytes 压缩任意二进制，返回裸 .wdw 二进制
func (c *Client) CompressBytes(raw []byte) ([]byte, error) {
	form := url.Values{}
	form.Set("clwb", string(raw))
	form.Set("lx", "1")
	form.Set("fhlx", "2")
	return c.postFormRaw(form, true)
}

// DecompressBytes 解压 .wdw 二进制，返回原文
func (c *Client) DecompressBytes(raw []byte) (string, error) {
	q := url.Values{}
	q.Set("lx", "2")
	q.Set("fhlx", "2")
	q.Set("key", c.Key)
	if c.DeviceID != "" {
		q.Set("device_id", c.DeviceID)
	}
	reqURL := c.Endpoint + "?" + q.Encode()
	req, err := http.NewRequest("POST", reqURL, bytes.NewReader(raw))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/octet-stream")
	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if err := checkError(body); err != nil {
		return "", err
	}
	return string(body), nil
}

// ---------- 内部 ----------

func (c *Client) baseForm() url.Values {
	f := url.Values{}
	f.Set("key", c.Key)
	if c.DeviceID != "" {
		f.Set("device_id", c.DeviceID)
	}
	return f
}

func (c *Client) postForm(form url.Values, binary bool) (string, error) {
	// 把 base 字段合并进去
	base := c.baseForm()
	for k, vs := range base {
		for _, v := range vs {
			form.Set(k, v)
		}
	}
	resp, err := c.HTTPClient.PostForm(c.Endpoint, form)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if err := checkError(body); err != nil {
		return "", err
	}
	return string(body), nil
}

func (c *Client) postFormRaw(form url.Values, binary bool) ([]byte, error) {
	base := c.baseForm()
	for k, vs := range base {
		for _, v := range vs {
			form.Set(k, v)
		}
	}
	resp, err := c.HTTPClient.PostForm(c.Endpoint, form)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if err := checkError(body); err != nil {
		return nil, err
	}
	return body, nil
}

// 失败时服务器返回 {"code":0,"msg":"..."}；成功时是裸文本/裸二进制
func checkError(body []byte) error {
	trimmed := bytes.TrimLeft(body, " \t\r\n")
	if len(trimmed) > 0 && trimmed[0] == '{' {
		var obj struct {
			Code int             `json:"code"`
			Msg  string          `json:"msg"`
			Raw  json.RawMessage `json:"data"`
		}
		if err := json.Unmarshal(trimmed, &obj); err == nil && obj.Msg != "" {
			return fmt.Errorf("wdw: %s", obj.Msg)
		}
	}
	return nil
}

// ---------- 本地 base85 <-> bytes (不消耗额度) ----------

const b85Alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz!#$%&()*+-;<=>?@^_`{|}~"

// B85Decode 把 <~...~> base85 文本解码为原始 bytes
func B85Decode(s string) ([]byte, error) {
	s = strings.TrimSpace(s)
	s = strings.TrimPrefix(s, "<~")
	s = strings.TrimSuffix(s, "~>")

	var val [256]byte
	for i := range val {
		val[i] = 255
	}
	for i := 0; i < len(b85Alphabet); i++ {
		val[b85Alphabet[i]] = byte(i)
	}

	var out []byte
	group := uint32(0)
	count := 0
	for i := 0; i < len(s); i++ {
		c := s[i]
		if val[c] == 255 {
			continue
		}
		group = group*85 + uint32(val[c])
		count++
		if count == 5 {
			out = append(out, byte(group>>24), byte(group>>16), byte(group>>8), byte(group))
			group = 0
			count = 0
		}
	}
	if count > 0 {
		for j := 0; j < 5-count; j++ {
			group *= 85
		}
		tmp := [4]byte{
			byte(group >> 24), byte(group >> 16), byte(group >> 8), byte(group),
		}
		for j := 0; j < count-1; j++ {
			out = append(out, tmp[j])
		}
	}
	return out, nil
}

// B85Encode 把任意 bytes 编码成 <~...~> 包装的 base85
func B85Encode(data []byte) string {
	var sb strings.Builder
	sb.WriteString("<~")
	for i := 0; i < len(data); i += 4 {
		chunk := data[i:min(i+4, len(data))]
		pad := 4 - len(chunk)
		var n uint32
		for _, b := range chunk {
			n = (n << 8) | uint32(b)
		}
		for j := len(chunk); j < 4; j++ {
			n <<= 8
		}
		codes := [5]int{}
		for d := 4; d >= 0; d-- {
			codes[d] = int(n % 85)
			n /= 85
		}
		if pad > 0 {
			keep := 5 - pad
			tailZero := true
			for d := keep; d < 5; d++ {
				if codes[d] != 0 { tailZero = false; break }
			}
			if !tailZero {
				pos := keep - 1
				codes[pos]++
				for pos > 0 && codes[pos] >= 85 {
					codes[pos] -= 85
					pos--
					codes[pos]++
				}
			}
			for d := 0; d < keep; d++ {
				sb.WriteByte(b85Alphabet[codes[d]])
			}
		} else {
			for d := 0; d < 5; d++ {
				sb.WriteByte(b85Alphabet[codes[d]])
			}
		}
	}
	sb.WriteString("~>")
	return sb.String()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
